### ISSUE TITLE
Handle nodes going permanently down during tablet migration

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#16527]
It can happen that a node is lost during tablet migration involving that node. Migration will be stuck, blocking topology state machine. To recover from this, the current procedure is for the admin to execute `nodetool removenode` or replacing the node. This act should first mark the node as down in group0 (on topology team: https://github.com/scylladb/scylladb/issues/16108) so that tablet state machine can pick this up and abort the migration.

This issue is about handling this in the tablet state machine.

There are the following cases:
   1) migration doesn't involve dead replica
   2) leaving replica is dead
   3) pending replica is dead
   4) both leaving and pending replicas are dead

First thing we should do is to make sure that the barrier ignores dead nodes, which will allow the migration to finish normally in the first case. It's also required for progress in other cases, which execute barriers, in addition to taking other actions described below. 

In the other cases (migration involves dead node), we either finish normally or revert to old replicas, depending on the stage we're in. 

To handle the decision to revert to old replicas, we add a new track of stages to which we can jump:
   1) cleanup_target - Coordinators use old replicas. The goal is to cleanup the pending replica, which may already received some writes. First execute the barrier. Then do the cleanup RPC to the pending replica, then jump to "revert_migration". If pending node is dead, the call to cleanup RPC is immediately satisfied.
   1) revert_migration - Coordinators use old replicas. Like end_migration, but reverts to old replicas instead of new replicas.

Handling of dead nodes during normal migration stages, when either leaving or pending (or both) are dead:
   1) allow_write_both_read_old - jump to a new stage called "revert_migration".
   2) write_both_read_old - jump to "cleanup_target".
   1) streaming - streaming will fail, and should not be retried. Jump to "cleanup_target" stage.
   1) write_both_read_new - We can either jump to the track which reverts to old replicas ("cleanup_target") or proceed. Which one to chose depends on which replica set has less number of dead nodes. 
   1) use_new - We cannot revert to old replicas from this stage onward. We proceed normally.
   1) cleanup - If leaving replica is dead, the cleanup RPC is immediately satisfied, and we can proceed.
   1) end_migration - Proceed normally.

The state machine logic can do the checks at the beginning of each stage handling (post barrier), and rely on the automatic retry mechanism to handle failures arising during streaming/RPCs, which will redirect coordinator back to the beginning of the stage handler.

Note that the above procedure deals with multiple nodes being dead too. Also, deals with nodes being marked as dead in the track which reverts to old replicas.

It may happen that all replicas are dead, and we'll finish the migration anyway, either by reverting to the old or the new replica set. This may result in permanent unavailability, but that will have to be dealt with by a different kind of transition, for example the one executed as part of nodetool removenode, which finds new tablet replicas.

When marking nodes as dead, we probably should at least warn if doing so would result in a loss of a quorum for some tablet. Should we require consent from admin in this case (`--force`)? 

Prerequisites:
- [x] https://github.com/scylladb/scylladb/issues/16108
- [ ] https://github.com/scylladb/scylladb/issues/16372

**Comment by kbr-scylla** [comment]
@tgrabiec please write a comment in https://github.com/scylladb/scylladb/issues/16108 and describe what you need precisely from raft-topology side, so we don't implement the wrong thing that turns out to be useless for your needs.

**Comment by tgrabiec** [comment]
> @tgrabiec is the 6.0 milestone realistic for this issue?

I put 6.0 for must-have items.

**Comment by bhalevy** [comment]
We hit a similar issue with the test_rolling_cluster_replacement_sequentially_dead_nodes_multi_dc dtest also when a dead node is replaced.
The following messages repeat on group0 leader log:
```
INFO  2024-01-28 20:59:33,971 [shard 0:strm] raft_topology - updating topology state: advance fence version to 44
INFO  2024-01-28 20:59:33,990 [shard 0:strm] raft_topology - executing global topology command barrier, excluded nodes: {}
ERROR 2024-01-28 20:59:34,018 [shard 0:strm] raft_topology - barrier for tablet cab2ecc0-be0d-11ee-a977-5371a1214ead:3 failed: seastar::broken_promise (broken promise)
ERROR 2024-01-28 20:59:34,018 [shard 0:strm] raft_topology - topology change coordinator fiber got error std::runtime_error (raft topology: exec_global_command(barrier) failed with seastar::rpc::closed_error (connection is closed))
INFO  2024-01-28 20:59:35,018 [shard 0:strm] raft_topology - entered `tablet migration` transition state
INFO  2024-01-28 20:59:35,018 [shard 0:strm] raft_topology - Initiating tablet cleanup of cab2ecc0-be0d-11ee-a977-5371a1214ead:1 on d40e2bdd-2332-46f1-aa1a-76a38ba03eae:1
INFO  2024-01-28 20:59:35,018 [shard 0:strm] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {}
ERROR 2024-01-28 20:59:35,019 [shard 0:strm] raft_topology - cleanup for tablet cab2ecc0-be0d-11ee-a977-5371a1214ead:1 failed: seastar::rpc::closed_error (connection is closed)
ERROR 2024-01-28 20:59:35,019 [shard 0:strm] raft_topology - drain rpc failed, proceed to fence old writes: std::runtime_error (raft topology: exec_global_command(barrier_and_drain) failed with seastar::rpc::closed_error (connection is closed))
```

@tgrabiec Does this deserve a new issue or can it be tracked here?

**Comment by tgrabiec** [comment]
> We hit a similar issue with the test_rolling_cluster_replacement_sequentially_dead_nodes_multi_dc dtest also when a dead node is replaced. The following messages repeat on group0 leader log:
> 
> ```
> INFO  2024-01-28 20:59:33,971 [shard 0:strm] raft_topology - updating topology state: advance fence version to 44
> INFO  2024-01-28 20:59:33,990 [shard 0:strm] raft_topology - executing global topology command barrier, excluded nodes: {}
> ERROR 2024-01-28 20:59:34,018 [shard 0:strm] raft_topology - barrier for tablet cab2ecc0-be0d-11ee-a977-5371a1214ead:3 failed: seastar::broken_promise (broken promise)
> ERROR 2024-01-28 20:59:34,018 [shard 0:strm] raft_topology - topology change coordinator fiber got error std::runtime_error (raft topology: exec_global_command(barrier) failed with seastar::rpc::closed_error (connection is closed))
> INFO  2024-01-28 20:59:35,018 [shard 0:strm] raft_topology - entered `tablet migration` transition state
> INFO  2024-01-28 20:59:35,018 [shard 0:strm] raft_topology - Initiating tablet cleanup of cab2ecc0-be0d-11ee-a977-5371a1214ead:1 on d40e2bdd-2332-46f1-aa1a-76a38ba03eae:1
> INFO  2024-01-28 20:59:35,018 [shard 0:strm] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {}
> ERROR 2024-01-28 20:59:35,019 [shard 0:strm] raft_topology - cleanup for tablet cab2ecc0-be0d-11ee-a977-5371a1214ead:1 failed: seastar::rpc::closed_error (connection is closed)
> ERROR 2024-01-28 20:59:35,019 [shard 0:strm] raft_topology - drain rpc failed, proceed to fence old writes: std::runtime_error (raft topology: exec_global_command(barrier_and_drain) failed with seastar::rpc::closed_error (connection is closed))
> ```
> 
> @tgrabiec Does this deserve a new issue or can it be tracked here?

It probably falls under this issue, if migrations started before node removal. To confirm, I need to look at the full log. Do you have a link?

**Comment by bhalevy** [comment]
>  To confirm, I need to look at the full log. Do you have a link?

No, becauss it ran locally. But since it failed every time it shouldn't be hard to reproduce on jankins with a byo run. I'll do that and post the link.

**Comment by tgrabiec** [comment]
The barrier already excludes dead nodes by consulting `get_excluded_nodes()`. Gleb's pending series makes this list permanent. Currently, it's only valid during topology change (e.g. during tablet draining on removenode).

```
    future<group0_guard> global_tablet_token_metadata_barrier(group0_guard guard) {
        // FIXME: Don't require all nodes to be up, only tablet replicas.
        return global_token_metadata_barrier(std::move(guard), _topo_sm._topology.get_excluded_nodes());
    }
```

**Comment by nyh** [comment]
The same problem that @bhalevy reported above also happens in the test `materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_duri`.

That test starts a cluster of four nodes, fills a base table, and then adds a view (which causes view building - and in particular writes - to happen in the background) while in parallel does a **removenode** of the second node. We get what seems like an endless repetitions of the messages, exactly one second apart:
```
INFO  2024-02-12 17:19:17,420 [shard 0:strm] raft_topology - entered `tablet migration` transition state
INFO  2024-02-12 17:19:17,420 [shard 0:strm] raft_topology - Initiating tablet cleanup of 8763b200-c9b7-11ee-92cb-f3c60238f2b0:1 on d49ce443-22b9-4679-add3-334490d90e21:0
INFO  2024-02-12 17:19:17,420 [shard 0:strm] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {}
ERROR 2024-02-12 17:19:17,420 [shard 0:strm] raft_topology - cleanup for tablet 8763b200-c9b7-11ee-92cb-f3c60238f2b0:1 failed: seastar::rpc::closed_error (connection is closed)
ERROR 2024-02-12 17:19:17,420 [shard 0:strm] raft_topology - drain rpc failed, proceed to fence old writes: std::runtime_error (raft topology: exec_global_command(barrier_and_drain) failed with seastar::rpc::closed_error (connection is closed))
INFO  2024-02-12 17:19:17,421 [shard 0:strm] raft_topology - updating topology state: advance fence version to 42
INFO  2024-02-12 17:19:17,429 [shard 0:strm] raft_topology - executing global topology command barrier, excluded nodes: {}
ERROR 2024-02-12 17:19:17,437 [shard 0:strm] raft_topology - barrier for tablet 8763b200-c9b7-11ee-92cb-f3c60238f2b0:0 failed: seastar::broken_promise (broken promise)
ERROR 2024-02-12 17:19:17,437 [shard 0:strm] raft_topology - barrier for tablet 876d7600-c9b7-11ee-92cb-f3c60238f2b0:1 failed: seastar::broken_promise (broken promise)
ERROR 2024-02-12 17:19:17,437 [shard 0:strm] raft_topology - barrier for tablet 87675b80-c9b7-11ee-92cb-f3c60238f2b0:3 failed: seastar::broken_promise (broken promise)
ERROR 2024-02-12 17:19:17,437 [shard 0:strm] raft_topology - barrier for tablet 87675b80-c9b7-11ee-92cb-f3c60238f2b0:2 failed: seastar::broken_promise (broken promise)
ERROR 2024-02-12 17:19:17,437 [shard 0:strm] raft_topology - barrier for tablet 87675b80-c9b7-11ee-92cb-f3c60238f2b0:0 failed: seastar::broken_promise (broken promise)
ERROR 2024-02-12 17:19:17,437 [shard 0:strm] raft_topology - barrier for tablet 81318b50-c9b7-11ee-92cb-f3c60238f2b0:0 failed: seastar::broken_promise (broken promise)
ERROR 2024-02-12 17:19:17,437 [shard 0:strm] raft_topology - topology change coordinator fiber got error std::runtime_error (raft topology: exec_global_command(barrier) failed with seastar::rpc::closed_error (connection is closed))
```

The test also never ends, so apparently these messages are not benign - something really stops working properly.

**Comment by tgrabiec** [comment]
@nyh You hit this issue, node went down during tablet migration, which blocks removenode which waits for tablet load balancing to release the topology state machine. The problem will be fixed when this issue is fixed.

**Comment by bhalevy** [comment]
We hit this issue with the update_cluster_layout_tests.py::TestUpdateClusterLayout::test_remove_node_from_gossip dtest.
I'll disable it for tablets for now.

**Comment by xemul** [comment]
There's a deadlock. Consider the following scenario.

Cluster with 4 nodes {A, B, C, D}
Tablet with replicas on {A, B}
Tablet migration B->C starts
Node B goes down while streaming
Operator calls removenode B // this marks B as ingored, but waits until removenode cycles through raft log
State machine switches to cleanup_target stage calling cleanup on C
Node C goes down while doing it
Operator calls removenode C, but this operation fails early since removenode B is in progress

The removenode marked B as ignored, but API call waits for the full removenode cycle to complete. It waits for raft log, raft log waits for cleanup_target, cleanup_target waits for C to be marked as ingored

@tgrabiec

**Comment by tgrabiec** [comment]
> The removenode marked B as ignored, but API call waits for the full removenode cycle to complete. It waits for raft log, raft log waits for cleanup_target, cleanup_target waits for C to be marked as ingored

I don't understand the problem description. What do you mean by "waits for raft log"? marking a node as ignored is a group0 commit, which shouldn't be blocked by topology operations like "cleanup_target".  But if we have 4 nodes and lost 2 of them, the we lose group0 quorum, maybe that's what's blocking group0 operations.

**Comment by tgrabiec** [comment]
If we lose group0 quorum, no topology operations can proceed. We need to execute a group0 recovery procedure.

**Comment by xemul** [comment]
> What do you mean by "waits for raft log"?

```
api - remove_node: host_id=ebaba1ce-0483-4f1d-89ac-dae954d09bcb ignore_nodes={}
raft_topology - request removenode for: ebaba1ce-0483-4f1d-89ac-dae954d09bcb, new ignored nodes: {ebaba1ce-0483-4f1d-89ac-dae954d09bcb}, existing ignore nodes: {}
...
raft_topology - removenode: wait for completion
```

and this "wait for completion" doesn't end, it loops around

```
raft_topology - entered `tablet migration` transition state
```

trying to complete the tablet migration

**Comment by xemul** [comment]
> Hm... Checking

No, the problem is elsewhere. I start with {A, B, C, D, E} nodes, initiate B->C migration, then stop-and-remove node B, then stop-and-remove node C, and removenode(C) aborts early (*) because removenode(B) is still running

(*) in `storage_service::run_with_api_lock("removenode")`

**Comment by xemul** [comment]
Seems to be purely API-level locking issue -- if calling removenode on some other host (say, D from above scenario) it processes eventually

**Comment by kbr-scylla** [comment]
> (*) in storage_service::run_with_api_lock("removenode")

We need to get rid of this in raft-topology mode.

**Comment by kbr-scylla** [comment]
cc @gleb-cloudius

**Comment by tgrabiec** [comment]
> and this "wait for completion" doesn't end, it loops around

It's expected that removenode as a whole waits for tablet migration to end. But you should still be able to execute another removenode for a different node, I think.

> Operator calls removenode C, but this operation fails early since removenode B is in progress

Where does it fail? Seems like it should at least manage to mark the node as dead. @gleb-cloudius @kbr-scylla

**Comment by xemul** [comment]
> It's expected that removenode as a whole waits for tablet migration to end. But you should still be able to execute another removenode for a different node, I think.

Yes, removenode via different node [works](https://github.com/scylladb/scylladb/issues/16527#issuecomment-1983417475)

> Where does it fail? Seems like it should at least manage to mark the node as dead.

In [storage service API](https://github.com/scylladb/scylladb/issues/16527#issuecomment-1983165921), and it does not manage to do anything with nodes

**Comment by tgrabiec** [comment]
> > Hm... Checking
> 
> No, the problem is elsewhere. I start with {A, B, C, D, E} nodes, initiate B->C migration, then stop-and-remove node B, then stop-and-remove node C, and removenode(C) aborts early (*) because removenode(B) is still running
> 
> (*) in `storage_service::run_with_api_lock("removenode")`

Please open an issue about this.

**Comment by tgrabiec** [comment]
@xemul I've noticed one issue with the already merged code:

```
                    if (tablet_state.streaming && tablet_state.streaming->failed()) {
                        if (check_excluded_replicas()) {
                            transition_to_with_barrier(locator::tablet_transition_stage::cleanup_target);
                            break;
                        }
                    }
```

When jumping to cleanup_target from streaming stage, we should also clear session `.del_session(last_token)`. If the leaving replica is dead, pending replica may still process some incoming mutations blocked in the pipeline, and that would race with the cleanup. Deleting session prior to executing barrier form the next stage ensures that such mutations will not be applied.

**Comment by yarongilor** [comment]
> We hit this issue with the update_cluster_layout_tests.py::TestUpdateClusterLayout::test_remove_node_from_gossip dtest. I'll disable it for tablets for now.

The test currently still fails with:
```
/home/yarongilor/.pyenv/versions/dtest-3.11.4/bin/python /home/yarongilor/.local/share/JetBrains/Toolbox/apps/PyCharm-C/ch-0/233.14475.56/plugins/python-ce/helpers/pycharm/_jb_pytest_runner.py --target update_cluster_layout_tests.py::TestUpdateClusterLayout.test_remove_node_from_gossip -- --scylla-version=unstable/master:2024-03-21T10:39:24Z --log-cli-level=debug --keep-test-dir --tablets 
Testing started at 13:09 ...
Launching pytest with arguments --scylla-version=unstable/master:2024-03-21T10:39:24Z --log-cli-level=debug --keep-test-dir --tablets update_cluster_layout_tests.py::TestUpdateClusterLayout::test_remove_node_from_gossip --no-header --no-summary -q in /home/yarongilor/git/scylla-dtest


---------------------------- live log sessionstart -----------------------------
13:09:47,509 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:48,479 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com?prefix=unstable%2Fscylla%2Fmaster%2Frelocatable%2F2024-03-21T10%3A39%3A24Z&encoding-type=url HTTP/1.1" 200 None
13:09:48,495 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:49,403 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10:39:24Z/00-Build.txt HTTP/1.1" 200 809
13:09:49,540 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:50,424 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "HEAD /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10%3A39%3A24Z/scylla-unified-5.5.0~dev-0.20240321.2b11842cb41b.x86_64.tar.gz HTTP/1.1" 200 0
============================= test session starts ==============================
collecting ... ----------------------------- live log collection ------------------------------
13:09:50,799 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:51,756 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com?prefix=unstable%2Fscylla%2Fmaster%2Frelocatable%2F2024-03-21T10%3A39%3A24Z&encoding-type=url HTTP/1.1" 200 None
13:09:51,779 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:52,766 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10:39:24Z/00-Build.txt HTTP/1.1" 200 809
13:09:52,903 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:53,910 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "HEAD /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10%3A39%3A24Z/scylla-unified-5.5.0~dev-0.20240321.2b11842cb41b.x86_64.tar.gz HTTP/1.1" 200 0
13:09:54,766 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:55,647 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com?prefix=unstable%2Fscylla%2Fmaster%2Frelocatable%2F2024-03-21T10%3A39%3A24Z&encoding-type=url HTTP/1.1" 200 None
13:09:55,653 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:56,508 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10:39:24Z/00-Build.txt HTTP/1.1" 200 809
13:09:56,558 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:57,490 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "HEAD /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10%3A39%3A24Z/scylla-unified-5.5.0~dev-0.20240321.2b11842cb41b.x86_64.tar.gz HTTP/1.1" 200 0
13:09:57,500 conftest        INFO  conftest.py    :100  | total available system memory is 31GB
13:09:57,501 conftest        DEBUG conftest.py    :439  | has sufficient resources? True
collected 1 item

update_cluster_layout_tests.py::TestUpdateClusterLayout::test_remove_node_from_gossip ------------------------------ live log logreport ------------------------------
13:20:43,664 errors          ERROR conftest.py    :201  | test failed: 
self = <update_cluster_layout_tests.TestUpdateClusterLayout object at 0x7f39dea9fcd0>

    def test_remove_node_from_gossip(self):
        """
        Test a node can be removed from gossip
        1. Create a cluster with 3 nodes
        2. Add node, wait for node to start bootstrappig
        3. Kill the new node
        4. Check that the new node will be removed from
        4. Check gossip on_remove callback in storage_service will not cause deadlock
        """
    
        # this error is expected in teardown after this test in raft topology mode
        ignore_error = r"raft_topology - send_raft_topology_cmd\(stream_ranges\) failed with exception \(node state is bootstrapping\)"
        self.ignore_log_patterns += [ignore_error]
    
        cluster = self.cluster
    
        # Disable hinted handoff and set batch commit log so this doesn't
        # interfere with the test (this must be after the populate)
        cluster.set_configuration_options(values=self.default_config_options(), batch_commitlog=True)
        cluster.populate(3).start()
        node1, node2, node3 = cluster.nodelist()
    
        session = self.patient_cql_connection(node1)
        create_ks(session, "ks", 2)
        create_cf(session, "cf", read_repair=0.0, columns={"c1": "text", "c2": "text"})
    
        insert_c1c2(session, keys=range(2000), consistency=ConsistencyLevel.TWO)
    
        logger.debug("Start node 4...")
        self.ignore_log_patterns += ["Startup failed"]
        node4 = new_node(cluster)
        node4.start(jvm_args=["--logger-log-level", "stream_session=debug"], no_wait=True)
        self.wait_for_node_streaming(node4)
    
        logger.debug("Hard-stop node 4 ...")
        node4.stop(gently=False)
    
        logger.debug("Check node 1 removed node4  ...")
        message = f"FatClient .*{node4.address()} has been silent for .*ms, removing from gossip"
        if "consistent-topology-changes" in self.scylla_features:
            message = f"Finished to force remove node {node4.address()}"
>       node1.watch_log_for(message)

update_cluster_layout_tests.py:1809: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ccmlib.scylla_node.ScyllaNode object at 0x7f39df1ee450>
exprs = 'Finished to force remove node 127.0.69.4', from_mark = None
timeout = 600, process = None, verbose = False, filename = 'system.log'

    def watch_log_for(self, exprs, from_mark=None, timeout=600, process=None, verbose=False, filename='system.log'):
        """
        Watch the log until one or more (regular) expression are found.
        This methods when all the expressions have been found or the method
        timeouts (a TimeoutError is then raised). On successful completion,
        a list of pair (line matched, match object) is returned.
        """
        elapsed = 0
        tofind = [exprs] if isinstance(exprs, str) else exprs
        tofind = [re.compile(e) for e in tofind]
        matchings = []
        reads = ""
        if len(tofind) == 0:
            return None
    
        log_file = os.path.join(self.get_path(), 'logs', filename)
        while not os.path.exists(log_file):
            time.sleep(.1)
            if process:
                process.poll()
                if process.returncode is not None:
                    self.print_process_output(self.name, process, verbose)
                    if process.returncode != 0:
                        raise RuntimeError()  # Shouldn't reuse RuntimeError but I'm lazy
    
        with open(log_file) as f:
            if from_mark:
                f.seek(from_mark)
    
            while True:
                # First, if we have a process to check, then check it.
                # Skip on Windows - stdout/stderr is cassandra.bat
                if not common.is_win():
                    if process:
                        process.poll()
                        if process.returncode is not None:
                            self.print_process_output(self.name, process, verbose)
                            if process.returncode != 0:
                                raise RuntimeError()  # Shouldn't reuse RuntimeError but I'm lazy
    
                line = f.readline()
                if line:
                    reads = reads + line
                    for e in tofind:
                        m = e.search(line)
                        if m:
                            matchings.append((line, m))
    
                            # This is bogus! - One should not modify the list from inside the loop which iterates on
                            # that list.
                            # However since we are going to break from the loop a few lines below that should be ok.
                            tofind.remove(e)
    
                            if len(tofind) == 0:
                                return matchings[0] if isinstance(exprs, str) else matchings
                            break
                else:
                    # yep, it's ugly
                    time.sleep(0.01)
                    elapsed = elapsed + 1
                    if elapsed > 100 * timeout:
>                       raise TimeoutError(time.strftime("%d %b %Y %H:%M:%S", time.gmtime()) + " [" + self.name + "] Missing: " + str(
                            [e.pattern for e in tofind]) + ":\n" + reads[:50] + f".....\nSee {filename} for remainder")
E                       ccmlib.node.TimeoutError: 21 Mar 2024 11:20:43 [node1] Missing: ['Finished to force remove node 127.0.69.4']:
E                       Scylla version 5.5.0~dev-0.20240321.2b11842cb41b w.....
E                       See system.log for remainder

../scylla-ccm/ccmlib/node.py:493: TimeoutError

------------------------------ live log teardown -------------------------------
13:20:46,774 dtest_setup     DEBUG dtest_setup.py :85   | Freeing cluster ID 69: link /home/yarongilor/.dtest/69


======================== 1 failed in 656.31s (0:10:56) =========================

-------------------------------- live log setup --------------------------------
13:09:57,905 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:58,781 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com?prefix=unstable%2Fscylla%2Fmaster%2Frelocatable%2F2024-03-21T10%3A39%3A24Z&encoding-type=url HTTP/1.1" 200 None
13:09:58,803 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:09:59,710 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10:39:24Z/00-Build.txt HTTP/1.1" 200 809
13:09:59,794 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:00,552 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "HEAD /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10%3A39%3A24Z/scylla-unified-5.5.0~dev-0.20240321.2b11842cb41b.x86_64.tar.gz HTTP/1.1" 200 0
13:10:00,655 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:01,417 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com?prefix=unstable%2Fscylla%2Fmaster%2Frelocatable%2F2024-03-21T10%3A39%3A24Z&encoding-type=url HTTP/1.1" 200 None
13:10:01,436 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:02,305 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10:39:24Z/00-Build.txt HTTP/1.1" 200 809
13:10:02,392 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:03,532 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "HEAD /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10%3A39%3A24Z/scylla-unified-5.5.0~dev-0.20240321.2b11842cb41b.x86_64.tar.gz HTTP/1.1" 200 0
13:10:03,762 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:04,713 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com?prefix=unstable%2Fscylla%2Fmaster%2Frelocatable%2F2024-03-21T10%3A39%3A24Z&encoding-type=url HTTP/1.1" 200 None
13:10:04,734 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:05,682 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10:39:24Z/00-Build.txt HTTP/1.1" 200 809
13:10:05,771 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:06,530 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "HEAD /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10%3A39%3A24Z/scylla-unified-5.5.0~dev-0.20240321.2b11842cb41b.x86_64.tar.gz HTTP/1.1" 200 0
13:10:06,547 urllib3.connectionpool DEBUG connectionpool.py:246  | Starting new HTTP connection (1): 169.254.169.254:80
13:10:09,607 conftest        INFO  conftest.py    :229  | Starting execution of test_remove_node_from_gossip at 2024-03-21 13:10:09.607346
13:10:09,610 dtest_setup     INFO  dtest_setup.py :858  | cluster ccm directory: /home/yarongilor/.dtest/dtest-9dozv334
13:10:09,611 ccm             DEBUG cluster.py     :762  | ScyllaCluster: cassandra_version=unstable/master:2024-03-21T10:39:24Z docker_image=None install_dir=None scylla_mode=None default_wait_other_notice_timeout=120 default_wait_for_binary_proto=420
13:10:09,703 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:10,571 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com?prefix=unstable%2Fscylla%2Fmaster%2Frelocatable%2F2024-03-21T10%3A39%3A24Z&encoding-type=url HTTP/1.1" 200 None
13:10:10,593 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:11,519 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "GET /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10:39:24Z/00-Build.txt HTTP/1.1" 200 809
13:10:11,604 urllib3.connectionpool DEBUG connectionpool.py:1019 | Starting new HTTPS connection (1): s3.amazonaws.com:443
13:10:12,533 urllib3.connectionpool DEBUG connectionpool.py:474  | https://s3.amazonaws.com:443 "HEAD /downloads.scylladb.com/unstable/scylla/master/relocatable/2024-03-21T10%3A39%3A24Z/scylla-unified-5.5.0~dev-0.20240321.2b11842cb41b.x86_64.tar.gz HTTP/1.1" 200 0
13:10:12,542 ccm             DEBUG cluster.py     :762  | ScyllaCluster: load_from_repository: install_dir=/home/yarongilor/.ccm/scylla-repository/unstable/master/2024-03-21T10_39_24Z scylla_mode=release default_wait_other_notice_timeout=120 default_wait_for_binary_proto=420
13:10:12,556 ccm             DEBUG cluster.py     :762  | Started cluster 'test' version 5.5.0-dev installed in /home/yarongilor/.ccm/scylla-repository/unstable/master/2024-03-21T10_39_24Z
13:10:12,557 dtest_setup     DEBUG dtest_setup.py :763  | Scylla mode is 'release'
13:10:12,558 dtest_setup     DEBUG dtest_setup.py :764  | Cluster *_request_timeout_in_ms=10000, range_request_timeout_in_ms=30000, cql request_timeout=30
13:10:12,566 dtest_setup     DEBUG dtest_setup.py :826  | Enabling tablets by setting experimental_features
13:10:12,575 dtest_setup     DEBUG dtest_setup.py :831  | Done setting configuration options:
{   'cas_contention_timeout_in_ms': 10000,
    'counter_write_request_timeout_in_ms': 20000,
    'experimental_features': ['tablets', 'consistent-topology-changes'],
    'initial_token': None,
    'num_tokens': 256,
    'phi_convict_threshold': 5,
    'range_request_timeout_in_ms': 30000,
    'read_request_timeout_in_ms': 10000,
    'request_timeout_in_ms': 10000,
    'ring_delay_ms': 10000,
    'skip_wait_for_gossip_to_settle': 0,
    'task_ttl_in_seconds': 0,
    'truncate_request_timeout_in_ms': 30000,
    'write_request_timeout_in_ms': 10000}
13:10:12,582 dtest_setup     DEBUG dtest_setup.py :61   | Allocated cluster ID 69: /home/yarongilor/.dtest/dtest-9dozv334
-------------------------------- live log call ---------------------------------
13:10:12,708 ccm             DEBUG cluster.py     :762  | start_nodes: no_wait=False wait_for_binary_proto=True wait_other_notice=True wait_normal_token_owner=True force_wait_for_cluster_start=True
13:10:12,711 ccm             DEBUG cluster.py     :762  | node1: Starting scylla: args=['/home/yarongilor/.dtest/dtest-9dozv334/test/node1/bin/scylla', '--options-file', '/home/yarongilor/.dtest/dtest-9dozv334/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', '127.0.69.1', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--overprovisioned', '--prometheus-address', '127.0.69.1', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
13:10:13,048 ccm             DEBUG cluster.py     :762  | node1: Starting scylla-jmx: args=['/home/yarongilor/.dtest/dtest-9dozv334/test/node1/bin/symlinks/scylla-jmx', '-Dapiaddress=127.0.69.1', '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder', '-Djava.rmi.server.hostname=127.0.69.1', '-Dcom.sun.management.jmxremote', '-Dcom.sun.management.jmxremote.host=127.0.69.1', '-Dcom.sun.management.jmxremote.port=7199', '-Dcom.sun.management.jmxremote.rmi.port=7199', '-Dcom.sun.management.jmxremote.local.only=false', '-Xmx256m', '-XX:+UseSerialGC', '-Dcom.sun.management.jmxremote.authenticate=false', '-Dcom.sun.management.jmxremote.ssl=false', '-jar', '/home/yarongilor/.dtest/dtest-9dozv334/test/node1/bin/scylla-jmx-1.0.jar']
13:10:13,358 ccm             DEBUG cluster.py     :762  | node2: Starting scylla: args=['/home/yarongilor/.dtest/dtest-9dozv334/test/node2/bin/scylla', '--options-file', '/home/yarongilor/.dtest/dtest-9dozv334/test/node2/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', '127.0.69.2', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--overprovisioned', '--prometheus-address', '127.0.69.2', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
13:10:17,110 ccm             DEBUG cluster.py     :762  | node2: Starting scylla-jmx: args=['/home/yarongilor/.dtest/dtest-9dozv334/test/node2/bin/symlinks/scylla-jmx', '-Dapiaddress=127.0.69.2', '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder', '-Djava.rmi.server.hostname=127.0.69.2', '-Dcom.sun.management.jmxremote', '-Dcom.sun.management.jmxremote.host=127.0.69.2', '-Dcom.sun.management.jmxremote.port=7199', '-Dcom.sun.management.jmxremote.rmi.port=7199', '-Dcom.sun.management.jmxremote.local.only=false', '-Xmx256m', '-XX:+UseSerialGC', '-Dcom.sun.management.jmxremote.authenticate=false', '-Dcom.sun.management.jmxremote.ssl=false', '-jar', '/home/yarongilor/.dtest/dtest-9dozv334/test/node2/bin/scylla-jmx-1.0.jar']
13:10:17,419 ccm             DEBUG cluster.py     :762  | node3: Starting scylla: args=['/home/yarongilor/.dtest/dtest-9dozv334/test/node3/bin/scylla', '--options-file', '/home/yarongilor/.dtest/dtest-9dozv334/test/node3/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', '127.0.69.3', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--overprovisioned', '--prometheus-address', '127.0.69.3', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
13:10:21,932 ccm             DEBUG cluster.py     :762  | node3: Starting scylla-jmx: args=['/home/yarongilor/.dtest/dtest-9dozv334/test/node3/bin/symlinks/scylla-jmx', '-Dapiaddress=127.0.69.3', '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder', '-Djava.rmi.server.hostname=127.0.69.3', '-Dcom.sun.management.jmxremote', '-Dcom.sun.management.jmxremote.host=127.0.69.3', '-Dcom.sun.management.jmxremote.port=7199', '-Dcom.sun.management.jmxremote.rmi.port=7199', '-Dcom.sun.management.jmxremote.local.only=false', '-Xmx256m', '-XX:+UseSerialGC', '-Dcom.sun.management.jmxremote.authenticate=false', '-Dcom.sun.management.jmxremote.ssl=false', '-jar', '/home/yarongilor/.dtest/dtest-9dozv334/test/node3/bin/scylla-jmx-1.0.jar']
13:10:22,330 cassandra.policies INFO  policies.py    :290  | Using datacenter 'datacenter1' for DCAwareRoundRobinPolicy (via host '127.0.69.1:9042'); if incorrect, please specify a local_dc to the constructor, or limit contact points to local cluster nodes
13:10:22,331 cassandra.cluster INFO  dtest_setup.py :496  | New Cassandra host <Host: 127.0.69.2:9042 datacenter1> discovered
13:10:22,331 cassandra.cluster INFO  dtest_setup.py :496  | New Cassandra host <Host: 127.0.69.3:9042 datacenter1> discovered
13:10:22,385 cassandra.protocol WARNING libevreactor.py:371  | Server warning: Using Replication Factor replication_factor=2 lower than the minimum_replication_factor_warn_threshold=3 is not recommended.
13:10:25,370 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:1796 | Start node 4...
13:10:25,426 ccm             DEBUG cluster.py     :762  | node4: Starting scylla: args=['/home/yarongilor/.dtest/dtest-9dozv334/test/node4/bin/scylla', '--options-file', '/home/yarongilor/.dtest/dtest-9dozv334/test/node4/conf/scylla.yaml', '--log-to-stdout', '1', '--logger-log-level', 'stream_session=debug', '--api-address', '127.0.69.4', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--overprovisioned', '--prometheus-address', '127.0.69.4', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=False wait_for_binary_proto=False
13:10:25,428 ccm             DEBUG cluster.py     :762  | node4: Starting scylla-jmx: args=['/home/yarongilor/.dtest/dtest-9dozv334/test/node4/bin/symlinks/scylla-jmx', '-Dapiaddress=127.0.69.4', '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder', '-Djava.rmi.server.hostname=127.0.69.4', '-Dcom.sun.management.jmxremote', '-Dcom.sun.management.jmxremote.host=127.0.69.4', '-Dcom.sun.management.jmxremote.port=7199', '-Dcom.sun.management.jmxremote.rmi.port=7199', '-Dcom.sun.management.jmxremote.local.only=false', '-Xmx256m', '-XX:+UseSerialGC', '-Dcom.sun.management.jmxremote.authenticate=false', '-Dcom.sun.management.jmxremote.ssl=false', '-jar', '/home/yarongilor/.dtest/dtest-9dozv334/test/node4/bin/scylla-jmx-1.0.jar']
13:10:30,077 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:512  | Bootstrap node4: started streaming/repair
13:10:30,077 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:1802 | Hard-stop node 4 ...
13:10:31,081 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:1805 | Check node 1 removed node4  ...
FAILED                                                                   [100%]
update_cluster_layout_tests.py:1767 (TestUpdateClusterLayout.test_remove_node_from_gossip)
self = <update_cluster_layout_tests.TestUpdateClusterLayout object at 0x7f39dea9fcd0>

    def test_remove_node_from_gossip(self):
        """
        Test a node can be removed from gossip
        1. Create a cluster with 3 nodes
        2. Add node, wait for node to start bootstrappig
        3. Kill the new node
        4. Check that the new node will be removed from
        4. Check gossip on_remove callback in storage_service will not cause deadlock
        """
    
        # this error is expected in teardown after this test in raft topology mode
        ignore_error = r"raft_topology - send_raft_topology_cmd\(stream_ranges\) failed with exception \(node state is bootstrapping\)"
        self.ignore_log_patterns += [ignore_error]
    
        cluster = self.cluster
    
        # Disable hinted handoff and set batch commit log so this doesn't
        # interfere with the test (this must be after the populate)
        cluster.set_configuration_options(values=self.default_config_options(), batch_commitlog=True)
        cluster.populate(3).start()
        node1, node2, node3 = cluster.nodelist()
    
        session = self.patient_cql_connection(node1)
        create_ks(session, "ks", 2)
        create_cf(session, "cf", read_repair=0.0, columns={"c1": "text", "c2": "text"})
    
        insert_c1c2(session, keys=range(2000), consistency=ConsistencyLevel.TWO)
    
        logger.debug("Start node 4...")
        self.ignore_log_patterns += ["Startup failed"]
        node4 = new_node(cluster)
        node4.start(jvm_args=["--logger-log-level", "stream_session=debug"], no_wait=True)
        self.wait_for_node_streaming(node4)
    
        logger.debug("Hard-stop node 4 ...")
        node4.stop(gently=False)
    
        logger.debug("Check node 1 removed node4  ...")
        message = f"FatClient .*{node4.address()} has been silent for .*ms, removing from gossip"
        if "consistent-topology-changes" in self.scylla_features:
            message = f"Finished to force remove node {node4.address()}"
>       node1.watch_log_for(message)

update_cluster_layout_tests.py:1809: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ccmlib.scylla_node.ScyllaNode object at 0x7f39df1ee450>
exprs = 'Finished to force remove node 127.0.69.4', from_mark = None
timeout = 600, process = None, verbose = False, filename = 'system.log'

    def watch_log_for(self, exprs, from_mark=None, timeout=600, process=None, verbose=False, filename='system.log'):
        """
        Watch the log until one or more (regular) expression are found.
        This methods when all the expressions have been found or the method
        timeouts (a TimeoutError is then raised). On successful completion,
        a list of pair (line matched, match object) is returned.
        """
        elapsed = 0
        tofind = [exprs] if isinstance(exprs, str) else exprs
        tofind = [re.compile(e) for e in tofind]
        matchings = []
        reads = ""
        if len(tofind) == 0:
            return None
    
        log_file = os.path.join(self.get_path(), 'logs', filename)
        while not os.path.exists(log_file):
            time.sleep(.1)
            if process:
                process.poll()
                if process.returncode is not None:
                    self.print_process_output(self.name, process, verbose)
                    if process.returncode != 0:
                        raise RuntimeError()  # Shouldn't reuse RuntimeError but I'm lazy
    
        with open(log_file) as f:
            if from_mark:
                f.seek(from_mark)
    
            while True:
                # First, if we have a process to check, then check it.
                # Skip on Windows - stdout/stderr is cassandra.bat
                if not common.is_win():
                    if process:
                        process.poll()
                        if process.returncode is not None:
                            self.print_process_output(self.name, process, verbose)
                            if process.returncode != 0:
                                raise RuntimeError()  # Shouldn't reuse RuntimeError but I'm lazy
    
                line = f.readline()
                if line:
                    reads = reads + line
                    for e in tofind:
                        m = e.search(line)
                        if m:
                            matchings.append((line, m))
    
                            # This is bogus! - One should not modify the list from inside the loop which iterates on
                            # that list.
                            # However since we are going to break from the loop a few lines below that should be ok.
                            tofind.remove(e)
    
                            if len(tofind) == 0:
                                return matchings[0] if isinstance(exprs, str) else matchings
                            break
                else:
                    # yep, it's ugly
                    time.sleep(0.01)
                    elapsed = elapsed + 1
                    if elapsed > 100 * timeout:
>                       raise TimeoutError(time.strftime("%d %b %Y %H:%M:%S", time.gmtime()) + " [" + self.name + "] Missing: " + str(
                            [e.pattern for e in tofind]) + ":\n" + reads[:50] + f".....\nSee {filename} for remainder")
E                       ccmlib.node.TimeoutError: 21 Mar 2024 11:20:43 [node1] Missing: ['Finished to force remove node 127.0.69.4']:
E                       Scylla version 5.5.0~dev-0.20240321.2b11842cb41b w.....
E                       See system.log for remainder

../scylla-ccm/ccmlib/node.py:493: TimeoutError

Process finished with exit code 1

```
